From f6925886f27a39eed2e43c5b7b6c342d00f7a970 Mon Sep 17 00:00:00 2001
From: kpcyrd <git@rxv.cc>
Date: Mon, 17 May 2021 02:01:08 +0200
Subject: [PATCH] Reproducible Builds: Do not record kernel version in binaries

---
 builder/FindGlobals.cmake | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/builder/FindGlobals.cmake b/builder/FindGlobals.cmake
index 6c96fc8e03..b13886c380 100644
--- a/builder/FindGlobals.cmake
+++ b/builder/FindGlobals.cmake
@@ -201,7 +201,7 @@ endif()
 if( MFX_SYSTEM )
   set( BUILD_INFO "${MFX_SYSTEM} ${CMAKE_SYSTEM_VERSION} | ${CMAKE_CXX_COMPILER_ID} ${CMAKE_CXX_COMPILER_VERSION}" )
 else()
-  set( BUILD_INFO "${CMAKE_SYSTEM} ${CMAKE_SYSTEM_VERSION} | ${CMAKE_CXX_COMPILER_ID} ${CMAKE_CXX_COMPILER_VERSION}" )
+  set( BUILD_INFO "${CMAKE_SYSTEM_NAME} | ${CMAKE_CXX_COMPILER_ID} ${CMAKE_CXX_COMPILER_VERSION}" )
 endif()
 
 if(UNIX AND MFX_GLIBC)
